server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportUnpromotableReplicaRefreshAction.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.admin.indices.refresh;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+
+public class TransportUnpromotableReplicaRefreshAction extends HandledTransportAction<
+    UnpromotableReplicaRefreshRequest,
+    ActionResponse.Empty> {
+    public static final String NAME = RefreshAction.NAME + "[u]";
+    public static final ActionType<ActionResponse.Empty> TYPE = new ActionType<>(NAME, in -> ActionResponse.Empty.INSTANCE);
+    private static final Logger logger = LogManager.getLogger(TransportUnpromotableReplicaRefreshAction.class);
+    private final ClusterService clusterService;
+    private final TransportService transportService;
+    private final IndicesService indicesService;
+    private final Client client;
+
+    @Inject
+    public TransportUnpromotableReplicaRefreshAction(
+        ClusterService clusterService,
+        TransportService transportService,
+        ActionFilters actionFilters,
+        IndicesService indicesService,
+        Client client
+    ) {
+        super(NAME, transportService, actionFilters, UnpromotableReplicaRefreshRequest::new);
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+        this.indicesService = indicesService;
+        this.client = client;
+    }
+
+    @Override
+    protected void doExecute(Task task, UnpromotableReplicaRefreshRequest request, ActionListener<ActionResponse.Empty> listener) {
+        ActionListener.completeWith(listener, () -> {
+            assert request.getSegmentGeneration() != Engine.RefreshResult.UNKNOWN_GENERATION;
+            IndexShard shard = indicesService.indexServiceSafe(request.getShardId().getIndex()).getShard(request.getShardId().id());
+            shard.waitForSegmentGeneration(request.getSegmentGeneration(), listener.map(l -> ActionResponse.Empty.INSTANCE));
+            return ActionResponse.Empty.INSTANCE;
+        });
```

Comment:
Hmm. I didn't see this on Zoom earlier, but this now looks wrong to me. By using `ActionListener#completeWith` we're not waiting for the shard to reach the right segment generation. I think we should say this:

```
        try {
            assert request.getSegmentGeneration() != Engine.RefreshResult.UNKNOWN_GENERATION;
            IndexShard shard = indicesService.indexServiceSafe(request.getShardId().getIndex()).getShard(request.getShardId().id());
            shard.waitForSegmentGeneration(request.getSegmentGeneration(), listener.map(l -> ActionResponse.Empty.INSTANCE));
        } catch (Exception e) {
            listener.onFailure(e);
        }
```

Is that what you had before? If so, we must have been getting stuck because the shard was not advancing to a new commit for some reason.

Reply:
I put the completeWith, but I see now it's wrong. I'll adopt your suggestion, thanks!